core/lib/server/pub.ts

```diff
@@ -1217,35 +1203,35 @@ export async function getPubsWithRelatedValuesAndChildren<
 					// we need to do this weird cast, because kysely does not support typing the selecting from a later CTE
 					// which is possible only in a with recursive query
 					.selectFrom("root_pubs_limited as p" as unknown as "pubs as p")
-					.leftJoin("pub_values as pv", "p.id", "pv.pubId")
-					.leftJoin("pub_fields", "pub_fields.id", "pv.fieldId")
-					.$if(Boolean(fieldSlugs), (qb) =>
-						qb.where("pub_fields.slug", "in", fieldSlugs!)
-					)
 					// maybe move this to root_pubs to save a join?
 					.leftJoin("PubsInStages", "p.id", "PubsInStages.pubId")
+					.$if(Boolean(withRelatedPubs), (qb) =>
+						qb
+							.leftJoin("pub_values as pv", (join) =>
+								join.on((eb) =>
+									eb.and([
+										eb("pv.pubId", "=", eb.ref("p.id")),
+										eb("pv.relatedPubId", "is not", null),
+									])
+								)
+							)
+							.select("pv.relatedPubId")
+					)
```

Comment:
this (and similarly in the `union` part) is the big change:

instead of joining ALL values up here before filtering and during the recursive lookup, we only join on the relatedPubId in order to traverse the relations. we then fetch the values separately at the end.